src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -17,51 +18,97 @@
 @dataclass
 class LockfileMetadata:
     invalidation_digest: str | None
+    valid_interpreter_constraints: InterpreterConstraints | None
+
+    def is_valid_for(
+        self,
+        expected_invalidation_digest: str | None,
+        user_interpreter_constraints: InterpreterConstraints,
+        interpreter_universe: Iterable[str],
+    ) -> bool:
+        """Returns True if this `LockfileMetadata` represents a lockfile that can be used in the
+        current execution context.
+
+        A lockfile can be used in the current execution context if `expected_invalidation_digest ==
+        invalidation_digest`, and if `user_interpreter_constraints` matches no interpreters
+        specified by `valid_interpreter_constraints`.
+        """
+
+        if self.invalidation_digest != expected_invalidation_digest:
+            return False
+
+        if self.valid_interpreter_constraints is None:
+            # This lockfile matches all interpreter constraints (TODO: check this)
+            return True
+
+        valid_versions = self.valid_interpreter_constraints.enumerate_python_versions(
+            interpreter_universe=interpreter_universe
+        )
+        user_versions = user_interpreter_constraints.enumerate_python_versions(
+            interpreter_universe=interpreter_universe
+        )
+
+        return user_versions.issubset(valid_versions)
 
 
 def calculate_invalidation_digest(
-    requirements: FrozenOrderedSet[str], interpreter_constraints: InterpreterConstraints
+    requirements: FrozenOrderedSet[str],
 ) -> str:
     """Returns an invalidation digest for the given requirements and interpreter constraints."""
```

Comment:
I've updated the docstring here, but to me, this is still an "invalidation" digest, we're just not invalidating based on the same things, or am I missing something here.

Reply:
Same logic as above comment:

> Maybe rename to requirements_digest? The digest is no longer sufficient to check for validity, and hex_digest might be misleading that it would be enough.

This is only the invalidation digest for requirements, not for the entire lockfile.